### PR TITLE
Feature/pyramid reproject map blocks

### LINKED ISSUE #10: Use xarray.map_blocks to speed up pyramid_reproject
@norlandrhagen and I have been experimenting with approaches for speeding up pyramid generation when using rio-xarray's `reproject` functionality. We have this rough prototype to share:

```python
def pyramid_reproject(
    ds, levels: int = None, pixels_per_tile=128, resampling="average", extra_dim=None
) -> dt.DataTree:
    from rasterio.transform import Affine
    from rasterio.warp import Resampling

    # multiscales spec
    save_kwargs = {"levels": levels, "pixels_per_tile": pixels_per_tile}
    attrs = {
        "multiscales": _multiscales_template(
            datasets=[{"path": str(i) for i in range(levels)}],
            type="reduce",
            method="pyramid_reproject",
            version=_get_version(),
            kwargs=save_kwargs,
        )
    }

    # set up pyramid
    root = xr.Dataset(attrs=attrs)
    pyramid = dt.DataTree(data_objects={"root": root})

    def make_template(da, dim, dst_transform, shape=None):

        template = xr.DataArray(
            data=dask.array.empty(shape, chunks=shape), dims=("y", "x"), attrs=da.attrs
        )
        template = make_grid_ds(template, dim, dst_transform)
        template.coords["spatial_ref"] = xr.DataArray(np.array(1.0))
        return template

    def reproject(da, shape=None, dst_transform=None, resampling="average"):
        return da.rio.reproject(
            "EPSG:3857",
            resampling=Resampling[resampling],
            shape=shape,
            transform=dst_transform,
        )

    for level in range(levels):
        lkey = str(level)
        dim = 2 ** level * pixels_per_tile

        dst_transform = Affine.translation(-20026376.39, 20048966.10) * Affine.scale(
            (20026376.39 * 2) / dim, -(20048966.10 * 2) / dim
        )

        pyramid[lkey] = xr.Dataset(attrs=ds.attrs)
        for k, da in ds.items():
            template = make_template(ds[k], dim, dst_transform, (dim, dim))
            pyramid[lkey].ds[k] = xr.map_blocks(
                reproject,
                da,
                kwargs=dict(shape=(dim, dim), dst_transform=dst_transform),
                template=template,
            )

    return pyramid
```

### PR DESCRIPTION
**Pull request by jhamman**
This PR adds a new feature to the `pyramid_reproject` function. Rather than iterating over extra dimensions in each DataArray, we now allow the DataArray to be chunked as a dask.array. Under the hood we use `xarray.map_blocks` to execute rioxarray's reproject function. 

- closes #10

### CODE REVIEW

File: ndpyramid/reproject.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        dim = 2 ** level * pixels_per_tile
+
+        dst_transform = Affine.translation(-20026376.39, 20048966.10) * Affine.scale(
+            (20026376.39 * 2) / dim, -(20048966.10 * 2) / dim
+        )
+
+        pyramid[lkey] = xr.Dataset(attrs=ds.attrs)
+        shape = (dim, dim)
+        for k, da in ds.items():
+            template_shape = (chunked_dim_sizes) + shape  # TODO: pick up here.
```
Comment (andersy005) [inline ndpyramid/reproject.py:119]: where is `chunked_dim_sizes` supposed to come from?

Comment (jhamman) [reply ndpyramid/reproject.py:119]: Good question. Clearly I didn't finish this part.

If I recall correctly, this is mean to support cases where non-spatial dims are chunked. The most common examples would be chunks along the `band` or `time` dimensions.

To do this, we need to construct a template that includes the correct shape for each block.

**Comment by andersy005** [comment]
pre-commit.ci autofix

**Comment by maxrjones** [comment]
Thanks for starting work on this Joe and Anderson! I'm planning to pick this feature up, but since the code base has changed a bunch it'll be easier to work in a fresh PR so I'll close this one now.

**Pull request closed without merging**